### ISSUE TITLE
(meta) Generate release notes instead of manually writing them

### ISSUE THREAD

**Issue by Kixunil** [lightningnetwork/lnd#6091]
### Problem

When contributing to LND it can happen that a PR has to be rebased multiple times just because some other PR was merged with release notes updates. It also causes situations [like this](https://github.com/lightningnetwork/lnd/blob/6d59a2c807abd4dc2613bde14240221a941fb2a7/docs/release-notes/release-notes-0.14.2.md) where a section can be duplicated.
Having to do two pushes (because PR number can't be predicted) is annoying and exhausts CI resources.

### Proposed solution

Have some way to specify the information to put in release notes without affecting other PRs or editing a single file. A popular way of doing it is [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) however I don't think that every single commit is worth mentioning in release notes. One way to resolve it is requiring to either specify `Release notes:` section in the commit message (presumably at the end) or special text `No release notes` (or something else). During release process a script would then walk the commits and extract the data and join them automatically. This would also take care of the list of authors.

**Comment by Roasbeef** [comment]
Github also seems to have a feature to generate a set of final release notes when you make a tag/release, presumably from the PRs themselves? Haven't tried it out though.

Here's a link to the docs: https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes

**Comment by Kixunil** [comment]
TIL, very interesting!

**Comment by naveensrinivasan** [comment]
+1   one of the biggest pain points of contributing to LND is fixing the release notes conflict.

**Comment by guggero** [comment]
I agree that the release notes present quite a bit of annoyance currently. Automating that process is definitely desirable, though perhaps not as easy as it might sound.
I played around with the auto-generated release notes feature of GitHub a bit: https://github.com/guggero/release-notes-test/releases/tag/v0.0.1

I'll need to run a few more tests, but I think this might be tricky for us to use, mainly because of the following reasons:
 - We often do minor releases from a branch where we cherry-pick PRs. The  PR merge commits are lost due to the rebase and I assume GitHub won't recognize them anymore (to be tested).
 - For release candidates we might create multiple releases for the same version. We need to make sure that the "this is a pre-release" checkbox tells GitHub to not diff this vs the last RC but the last full release.
 - We like to have a bit more information in the release notes than just the PR title. Though maybe we need to sacrifice that in order to gain more automation.

**Comment by naveensrinivasan** [comment]
>We like to have a bit more information in the release notes than just the PR title. Though maybe we need to sacrifice that in order to gain more automation.

I agree, PR Titles aren't clear to have consistent release notes. 

Here is an example of trying to address this, though it might to be perfect https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process and an workflow to check if the PR follows the specific title https://github.com/ossf/scorecard/blob/main/.github/workflows/verify.yml

**Comment by Kixunil** [comment]
Found [git-cliff](https://github.com/orhun/git-cliff) today by chance, looks pretty good IMO.

**Comment by Kixunil** [comment]
Did some additional research and ultimately, I'm not content with any solution out there. I really, really like the current format of release notes and the only problematic thing here is merge conflicts. Specifically I like that a separate, user-facing description can be specified, not same as commit message and that things are neatly organized into sections. (I often like to check breaking changes first for instance.)

Given this situation [I have proposed to create our own tool](https://gist.github.com/Kixunil/f796f65500c4ca18ca629daaf2eca272). TL;DR is that I have someone who may be able to work on it and I tried to design it to fit current LND format and just moved inputs to commit messages. Being outputted into a file it allows for a quick manual check or modifications before committing.

Please let me know if this is interesting so that I can better decide if we should do it.

**Comment by guggero** [comment]
> I really, really like the current format of release notes

Cool, glad to hear!

Thank you for proposing the tool. I think you are correct, no current solutions out there (that I was able to find) are currently sufficient for what we would like to have. What you propose sounds great, I think building it should help a lot of projects.

One comment though: Wouldn't it instead be great if the annotations wouldn't be _per commit_ but _per PR_ instead? Otherwise, how would a multi-commit PR look like? Would the `changelog: skip` annotation be added to each but the last commit and that would then contain the full description? Shouldn't the default be `skip` and then we can have a CI pipeline that checks that at least one commit has the `changelog` annotation?

**Comment by Kixunil** [comment]
My intention was to go with all commits have to add information or skip which minimizes risk of forgetting but I guess it could be too much. Maybe there could be some additional filter that allows e.g. `test: ...` commits to be skipped by default.

I'm not opposed to having an option to do what you propose. I think it'll be quite easy to write anyway. Was also thinking about putting it in merge commits but IDK how it would interact with your merging workflow.

**Comment by guggero** [comment]
Forcing things to be in ever commit will be hard for us I think. There are a lot of one-time contributors that sometimes only use the GitHub UI to create PRs, so even editing the commit message after the fact seems to be a hassle.
If we could just have an annotation in the PR body, that would be perfect for our use case.
That could even be checked by a CI action, I think.

Merge commits also aren't ideal since we create them, not the contributor. And we want the contributors to write the release notes entry.

**Comment by Kixunil** [comment]
Hmm, I originally didn't want to complicate it by using GitHub API and such. Would it be feasible to copy-paste the changelog entry from PR to merge commit during merging?
Anyway, I will think about it more since your argument sounds compelling.

**Comment by guggero** [comment]
Hmm, you're right. Copy/pasting into the merge commit could be an option indeed.

**Comment by Kixunil** [comment]
Something will still be useful to avoid forgetting to copy-paste. One possibility is to have another command or tool that does both merging and copying.
